nexus/src/app/sagas/instance_migrate.rs

```diff
@@ -72,19 +70,27 @@ declare_saga_actions! {
 
     CREATE_MIGRATION_RECORD -> "migration_record" {
         + sim_create_migration_record
-        - sim_delete_migration_record
+        - sim_fail_migration_record
     }
 
-
     // This step the instance's migration ID and destination Propolis ID
-    // fields. Because the instance is active, its current sled agent maintains
-    // its most recent runtime state, so to update it, the saga calls into the
-    // sled and asks it to produce an updated instance record with the
-    // appropriate migration IDs and a new generation number.
+    // fields in the database.
     //
-    // The source sled agent synchronizes concurrent attempts to set these IDs.
-    // Setting a new migration ID and re-setting an existing ID are allowed, but
-    // trying to set an ID when a different ID is already present fails.
+    // If the instance's migration ID has already been set when we attempt to
+    // set ours, that means we have probably raced with another migrate saga for
+    // the same instance. If this is the case, this action will fail and the
+    // saga will unwind.
+    //
+    // Yes, it's a bit unfortunate that our attempt to compare-and-swap in a
+    // migration ID happens only after we've created VMM and migration records,
+    // and that we'll have to destroy them as we unwind. However, the
+    // alternative, setting the migration IDs *before* records for the target
+    // VMM and the migration are created, would mean that there is a period of
+    // time during which the instance record contains foreign keys into the
+    // `vmm` and `migration` tables that don't have corresponding records to
+    // those tables. Because the `instance` table is queried in the public API,
+    // we take care to ensure that it doesn't have "dangling pointers" to
+    // records in the `vmm` and `migration` tables that don't exist yet.
```

Comment:
Note that setting migration IDs for an instance is now just a database op in Nexus, and not a sled-agent API call.